types/recompose/index.d.ts

```diff
@@ -268,20 +269,22 @@ declare module 'recompose' {
 
     // Static property helpers: https://github.com/acdlite/recompose/blob/master/docs/API.md#static-property-helpers
 
+    type ComponentModifier = <T extends Component>(component: T) => T;
+
     // setStatic: https://github.com/acdlite/recompose/blob/master/docs/API.md#setStatic
-    export function setStatic<TOutter>(
+    export function setStatic(
         key: string, value: any
-    ): ComponentEnhancer<TOutter, TOutter>;
+    ): ComponentModifier;
 
     // setPropTypes: https://github.com/acdlite/recompose/blob/master/docs/API.md#setPropTypes
-    export function setPropTypes<TOutter>(
-        propTypes: ValidationMap<TOutter>
-    ): ComponentEnhancer<any, TOutter>;
+    export function setPropTypes<P>(
+        propTypes: ValidationMap<P>
+    ): <T extends Component<P>>(component: T) => T;
```

Comment:
If `ComponentModifier` were generic, could this be simplified?

Reply:
(i.e. as `ComponentModifier<P>`)

Reply:
`setPropTypes` is the only static property helper that needs to know the type of the props for the component that the resulting function will be called on, and is the only static property helper that should return a function that is limited to only being callable on a component with props of that type.

It needs to know the props since it is setting the `propTypes` static property which is [a validation map for the properties of the component](https://reactjs.org/docs/typechecking-with-proptypes.html).

In contrast, `setStatic` and `setDisplayName` both create functions that can be called on *any* component.

Making `ComponentModifier` generic would force `setStatic` and `setDisplayName` to require a type argument which would make the typing on the functions they return incorrect (more info below on the section for `setStatic`).

Reply:
If `ComponentModifier` had an optional type parameter (e.g. `ComponentModifier<P = {}>`), then `setStatic` and `setDisplayName` would not need to be generic, would they?  Or were you making a subtler objection (e.g. that doing it that way might somehow degrade type inference)?

Reply:
I played with that...if `ComponentModifier` has an optional type parameter then `setStatic` and `setDisplayName` still need to be generic and need to pass the correct `P` to `ComponentModifier`, if they don't then `P` will default to `{}` and the resulting function will be incorrectly typed as only callable on a component with no props.

Code update
---

I removed `ComponentModifier` since it seemed to be causing confusion and updated the typing for `setStatic` and `setDisplayName` to very clearly return a function that can be called on a component and will return what it was called on since that is what they do: [`setStatic`](https://github.com/acdlite/recompose/blob/68c560b216f2530796147bee07f45bf2b9bf0412/src/packages/recompose/setStatic.js#L3-L5), [`setDisplayName`](https://github.com/acdlite/recompose/blob/68c560b216f2530796147bee07f45bf2b9bf0412/src/packages/recompose/setDisplayName.js#L3).